diff -ru a/pyggi/tree/__init__.py b/pyggi/tree/__init__.py
--- a/pyggi/pyggi/tree/__init__.py	2020-05-14 13:19:54.000000000 +0100
+++ b/pyggi/pyggi/tree/__init__.py	2020-05-14 13:36:53.673436652 +0100
@@ -1,5 +1,5 @@
 from .abstract_engine import AbstractTreeEngine
-from .astor_engine import AstorEngine
+# from .astor_engine import AstorEngine
 from .xml_engine import XmlEngine
 from .tree import StmtReplacement, StmtInsertion, StmtDeletion, StmtMoving
 from .tree import TreeProgram
diff -ru a/pyggi/tree/tree.py b/pyggi/tree/tree.py
--- a/pyggi/pyggi/tree/tree.py	2020-05-14 13:19:54.000000000 +0100
+++ b/pyggi/pyggi/tree/tree.py	2020-05-14 13:37:16.860103726 +0100
@@ -3,7 +3,8 @@
 import astor
 import random
 from abc import abstractmethod
-from . import AstorEngine, XmlEngine
+# from . import AstorEngine, XmlEngine
+from . import XmlEngine
 from ..base import AbstractProgram, AbstractEdit
 from ..utils import get_file_extension
 
@@ -11,9 +12,10 @@
     @classmethod
     def get_engine(cls, file_name):
         extension = get_file_extension(file_name)
-        if extension in ['.py']:
-            return AstorEngine
-        elif extension in ['.xml']:
+        # if extension in ['.py']:
+        #     return AstorEngine
+        # elif extension in ['.xml']:
+        if extension in ['.xml']:
             return XmlEngine
         else:
             raise Exception('{} file is not supported'.format(extension))
